src/Data/Vector/Hashtables/Internal.hs

```diff
@@ -106,15 +106,20 @@ findElem FrozenDictionary{..} key' = go $ fbuckets !. (hashCode' `rem` A.sizeofP
 
 -- | Infix version of @unsafeRead@.
 (!~) :: (MVector v a, PrimMonad m) => v (PrimState m) a -> Int -> m a
-(!~) = V.unsafeRead
+(!~) xs !i = V.unsafeRead xs i
+-- Why do we need ! before i?
+-- The reason is that V.unsafeRead is essentially V.basicUnsafeRead,
+-- which is an opaque class member and, unless V.unsafeRead was
+-- already specialised to a specific v, GHC has no clue that i is most certainly
+-- to be used eagerly. Bang before i hints this vital for optimizer information.
 
 -- | Infix version of @unsafeIndex@.
 (!.~) :: (Vector v a) => v a -> Int -> a
-(!.~) = VI.unsafeIndex
+(!.~) xs !i = VI.unsafeIndex xs i
```

Comment:
I don't think adding a bang here or below will ever change anything with the current implementation. One may argue they the implementation can, in theory, change.

Reply:
`VI.unsafeIndex` is a thin wrapper over `basicUnsafeIndexM` and GHC cannot figure out its strictness (it's an opaque class method), unless we hint it explicitly.

Reply:
Oh, sorry, I mixed up some things, scratch it.